apps/emqx_bridge_snowflake/src/emqx_bridge_snowflake_connector.erl

```diff
@@ -0,0 +1,944 @@
+        seq_no := SeqNo,
+        work_dir := WorkDir
+    } = TransferState,
+    Filename0 = emqx_template:render_strict(FilenameTemplate, #{
+        action_res_id => ActionResId,
+        seq_no => SeqNo,
+        container_type => ContainerType
+    }),
+    Filename = filename:join([WorkDir, <<"tmp">>, Filename0]),
+    ok = filelib:ensure_dir(Filename),
+    {ok, FD} = file:open(Filename, [write, binary]),
+    TransferState#{
+        filename := Filename,
+        fd := FD
+    };
+ensure_file(TransferState) ->
+    TransferState.
+
+append_to_file(IOData, Size, TransferState) ->
+    #{
+        fd := FD,
+        written := Written
+    } = TransferState,
+    %% Todo: handle errors?
+    ok = file:write(FD, IOData),
+    TransferState#{written := Written + Size}.
+
+close_and_enqueue_file(TransferState0) ->
+    #{
+        fd := FD,
+        filename := Filename,
+        next_file := NextFile,
+        seq_no := SeqNo,
+        written := Written
+    } = TransferState0,
+    ok = file:close(FD),
+    TransferState0#{
+        next_file := queue:in({Filename, Written}, NextFile),
+        filename := undefined,
+        fd := undefined,
+        seq_no := SeqNo + 1,
+        written := 0
+    }.
+
+-spec process_write(transfer_state()) ->
+    {ok, transfer_state()} | {error, term()}.
+process_write(TransferState0) ->
+    #{next_file := NextFile0} = TransferState0,
+    case queue:out(NextFile0) of
+        {{value, {Filename, Size}}, NextFile} ->
+            ?tp(snowflake_will_stage_file, #{}),
+            do_process_write(Filename, Size, TransferState0#{next_file := NextFile});
+        {empty, _} ->
+            {ok, TransferState0}
+    end.
+
+-spec do_process_write(file:filename(), non_neg_integer(), transfer_state()) ->
+    {ok, transfer_state()} | {error, term()}.
+do_process_write(Filename, Size, TransferState0) ->
+    #{
+        odbc_pool := ODBCPool,
+        database := Database,
+        schema := Schema,
+        stage := Stage,
+        staged_files := StagedFiles0
+    } = TransferState0,
+    case stage_file(ODBCPool, Filename, Database, Schema, Stage) of
+        {ok, Target} ->
+            StagedFile = #{path => Target, size => Size},
+            StagedFiles = [StagedFile | StagedFiles0],
+            TransferState = TransferState0#{staged_files := StagedFiles},
+            process_write(TransferState);
+        {error, Reason} ->
+            %% TODO: retry?
+            {error, Reason}
+    end.
+
+-spec process_complete(transfer_state()) ->
+    {ok, term()}.
+process_complete(TransferState0) ->
+    #{written := Written0} = TransferState0,
+    maybe
+        %% Flush any left-over data
+        {ok, TransferState} ?=
+            case Written0 > 0 of
+                true ->
+                    ?tp("snowflake_flush_on_complete", #{}),
+                    TransferState1 = close_and_enqueue_file(TransferState0),
+                    process_write(TransferState1);
```

Comment:
TODO: seems to be a bug where this attempts to write the same file twice after it was already skipped, specially when restarting the node.